### PR TITLE
Improve TMaster return message when metrics is not available

### LINKED ISSUE #1322: Unknown component error in Heron Tracker
```heron-tracker
INFO: 08/30/2016 22:54:25 -0700 Connecting to file state with rootpath: /Users/karthikz/.herondata/repository/state/local
INFO: 08/30/2016 22:54:25 -0700 Tracker has started
INFO: 08/30/2016 22:54:25 -0700 Running on port: 8888
INFO: 08/30/2016 22:54:25 -0700 Using config file: /Users/karthikz/.herontools/conf/heron_tracker.yaml
INFO: 08/30/2016 22:55:33 -0700 200 GET /clusters (::1) 0.67ms
INFO: 08/30/2016 22:55:33 -0700 200 GET /topologies/states (::1) 0.69ms
INFO: 08/30/2016 22:56:15 -0700 State watch triggered for topologies.
INFO: 08/30/2016 22:56:15 -0700 Adding new topology: AckingTopology, state_manager: local
INFO: 08/30/2016 22:56:15 -0700 Registering a watch with uid: e63d151e-0c66-484e-b769-fba9f2f47858
INFO: 08/30/2016 22:56:15 -0700 No execution state found for: AckingTopology
INFO: 08/30/2016 22:56:15 -0700 Watch triggered for topology execution state: AckingTopology
INFO: 08/30/2016 22:56:15 -0700 Setting topology info for topology: AckingTopology
INFO: 08/30/2016 22:56:15 -0700 Watch triggered for topology pplan: AckingTopology
INFO: 08/30/2016 22:56:15 -0700 Setting topology info for topology: AckingTopology
INFO: 08/30/2016 22:56:15 -0700 Watch triggered for topology tmaster: AckingTopology
INFO: 08/30/2016 22:56:15 -0700 Setting topology info for topology: AckingTopology
INFO: 08/30/2016 22:56:15 -0700 Watch triggered for topology scheduler location: AckingTopology
INFO: 08/30/2016 22:56:16 -0700 Setting topology info for topology: AckingTopology
INFO: 08/30/2016 22:56:21 -0700 Watch triggered for topology pplan: AckingTopology
INFO: 08/30/2016 22:56:21 -0700 Setting topology info for topology: AckingTopology
INFO: 08/30/2016 22:56:21 -0700 Watch triggered for topology tmaster: AckingTopology
INFO: 08/30/2016 22:56:21 -0700 Setting topology info for topology: AckingTopology
INFO: 08/30/2016 22:56:24 -0700 200 GET /clusters (::1) 0.39ms
INFO: 08/30/2016 22:56:24 -0700 200 GET /topologies/states (::1) 1.48ms
INFO: 08/30/2016 22:56:31 -0700 200 GET /topologies/executionstate?cluster=local&environ=default&topology=AckingTopology (::1) 7.37ms
INFO: 08/30/2016 22:56:31 -0700 200 GET /topologies/schedulerlocation?cluster=local&environ=default&topology=AckingTopology (::1) 2.43ms
INFO: 08/30/2016 22:56:31 -0700 200 GET /topologies/logicalplan?cluster=local&environ=default&topology=AckingTopology (::1) 0.46ms
INFO: 08/30/2016 22:56:31 -0700 200 GET /topologies/logicalplan?cluster=local&environ=default&topology=AckingTopology (::1) 0.74ms
INFO: 08/30/2016 22:56:31 -0700 200 GET /topologies/physicalplan?cluster=local&environ=default&topology=AckingTopology (::1) 0.46ms
INFO: 08/30/2016 22:56:31 -0700 200 GET /topologies/logicalplan?cluster=local&environ=default&topology=AckingTopology (::1) 0.40ms
INFO: 08/30/2016 22:56:31 -0700 200 GET /topologies/physicalplan?cluster=local&environ=default&topology=AckingTopology (::1) 0.45ms
INFO: 08/30/2016 22:56:31 -0700 200 GET /topologies/physicalplan?cluster=local&environ=default&topology=AckingTopology (::1) 0.60ms
ERROR: 08/30/2016 22:56:32 -0700 Unknown component
ERROR: 08/30/2016 22:56:32 -0700 Unknown component
ERROR: 08/30/2016 22:56:32 -0700 Unknown component
ERROR: 08/30/2016 22:56:32 -0700 Unknown component
ERROR: 08/30/2016 22:56:32 -0700 Unknown component
ERROR: 08/30/2016 22:56:32 -0700 Unknown component
INFO: 08/30/2016 22:56:32 -0700 200 GET /topologies/metrics?component=word&cluster=local&environ=default&topology=AckingTopology&metricname=__emit-count%2Fdefault&interval=600 (::1) 471.32ms
INFO: 08/30/2016 22:56:32 -0700 200 GET /topologies/metrics?component=word&cluster=local&environ=default&topology=AckingTopology&metricname=__complete-latency%2Fdefault&interval=3600 (::1) 26.73ms
INFO: 08/30/2016 22:56:32 -0700 200 GET /topologies/metrics?component=word&cluster=local&environ=default&topology=AckingTopology&metricname=__complete-latency%2Fdefault&interval=600 (::1) 23.34ms
INFO: 08/30/2016 22:56:32 -0700 200 GET /topologies/metrics?component=word&cluster=local&environ=default&topology=AckingTopology&metricname=__emit-count%2Fdefault&interval=-1 (::1) 19.68ms
INFO: 08/30/2016 22:56:32 -0700 200 GET /topologies/metrics?component=word&cluster=local&environ=default&topology=AckingTopology&metricname=__emit-count%2Fdefault&interval=3600 (::1) 16.19ms
INFO: 08/30/2016 22:56:32 -0700 200 GET /topologies/metrics?component=word&cluster=local&environ=default&topology=AckingTopology&metricname=__emit-count%2Fdefault&interval=10800 (::1) 16.90ms
ERROR: 08/30/2016 22:56:32 -0700 Unknown component
ERROR: 08/30/2016 22:56:32 -0700 Unknown component
ERROR: 08/30/2016 22:56:32 -0700 Unknown component
ERROR: 08/30/2016 22:56:32 -0700 Unknown component
ERROR: 08/30/2016 22:56:32 -0700 Unknown component
ERROR: 08/30/2016 22:56:32 -0700 Unknown component
```

### PR DESCRIPTION
**Pull request by objmagic**
In `GetMetrics` function, if there is no metrics associated with a component, we will now check if the component actually exists in the topology. If it exists in the topology, we know that the metrics is not yet available. We can then return a more proper error message.

resolves #1322 

cc @kramasamy @cckellogg @billonahill

### CODE REVIEW

**Comment by objmagic** [comment]
Not sure if I should change `Log.error` to `Log.warn` here: https://github.com/twitter/heron/pull/1336/files#diff-409c349551fd1e7ad3611d8544571cffR148

In my opinion, we should have different logging level for "Metrics not available" (`Log.warn`) and "Unknown component" (`Log.error`). However, we have no way to distinguish since we only have [`OK` and `NOTOK`](https://github.com/twitter/heron/blob/62793b63413976cc4f9d2e7cceb0815d062523d4/heron/proto/common.proto#L6). Maybe we can add another specific error number, like `METRICS_NOT_AVAILABLE`?

File: heron/tools/tracker/src/python/handlers/metricshandler.py [hunk 1]
```diff
@@ -145,7 +145,7 @@ def getComponentMetrics(self,
 
     if metricResponse.status.status == common_pb2.NOTOK:
       if metricResponse.status.HasField("message"):
-        Log.error(metricResponse.status.message)
+        Log.warn(metricResponse.status.message)
```
Comment (billonahill) [inline heron/tools/tracker/src/python/handlers/metricshandler.py]: Log.warn seems fine here, but I typically like to prepend some text to make it clear that this message is from the metrics manager. "Received the following response from the metrics manager: ..."

Same for metricstimeline.py below.

File: heron/tmaster/src/cpp/manager/tmetrics-collector.cpp [hunk 2]
```diff
@@ -104,12 +105,34 @@ void TMetricsCollector::AddMetric(const PublishMetrics& _metrics) {
...
+        component_exists = true;
+      }
+    }
+    for (int i = 0; i < _topology->bolts_size(); i++) {
+      if ((_topology->bolts(i)).comp().name() == _request.component_name()) {
+        component_exists = true;
+      }
+    }
+    if (component_exists) {
+      LOG(WARNING) << "Metrics for " << _request.component_name() << " is not available";
```
Comment (billonahill) [inline heron/tmaster/src/cpp/manager/tmetrics-collector.cpp]: _are_ not available

File: heron/tools/tracker/src/python/handlers/metricshandler.py (hunk 1 above)
Comment (objmagic) [inline heron/tools/tracker/src/python/handlers/metricshandler.py]: fixed via: https://github.com/twitter/heron/pull/1336/commits/18b6dd1011f8d5234686afe4814b837378595afd#diff-409c349551fd1e7ad3611d8544571cffR148

File: heron/tmaster/src/cpp/manager/tmetrics-collector.cpp (hunk 2 above)
Comment (objmagic) [inline heron/tmaster/src/cpp/manager/tmetrics-collector.cpp]: Thanks!!!

fixed via https://github.com/twitter/heron/pull/1336/commits/18b6dd1011f8d5234686afe4814b837378595afd#diff-b5454de1597b4cda2547ba9c6ca12541R125

**Comment by billonahill** [comment]
👍

**Comment by objmagic** [comment]
will merge this when internal integration test passes

File: heron/tmaster/src/cpp/manager/tmetrics-collector.cpp [hunk 3]
```diff
@@ -104,12 +106,34 @@ void TMetricsCollector::AddMetric(const PublishMetrics& _metrics) {
...
   MetricResponse* response = new MetricResponse();
+
   if (metrics_.find(_request.component_name()) == metrics_.end()) {
-    LOG(ERROR) << "GetMetrics request received for unknown component " << _request.component_name();
-    response->mutable_status()->set_status(proto::system::NOTOK);
-    response->mutable_status()->set_message("Unknown component");
+    bool component_exists = false;
+    for (int i = 0; i < _topology->spouts_size(); i++) {
+      if ((_topology->spouts(i)).comp().name() == _request.component_name()) {
+        component_exists = true;
```
Comment (congwang) [inline heron/tmaster/src/cpp/manager/tmetrics-collector.cpp]: I think you can break this loop and skip the following loop once find it.

Comment (objmagic) [inline heron/tmaster/src/cpp/manager/tmetrics-collector.cpp]: right.

Comment (objmagic) [inline heron/tmaster/src/cpp/manager/tmetrics-collector.cpp]: update via https://github.com/twitter/heron/pull/1336/commits/8ddd534675296dfe2032edd6bbf7b0d2c5650bc8. Can you have a look?

**Comment by kramasamy** [comment]
👍

**Pull request merged**